src/index.js

```diff
@@ -16,102 +14,151 @@ function makeDnsPromise(host) {
     endTime = Math.max(
       0,
-      context.getRemainingTimeInMillis() - config.timeoutWindow
+      modifiedContext.getRemainingTimeInMillis() - config.timeoutWindow
     );
   }
 
   return setTimeout(() => {
-    report.send(new Error('Timeout Exceeded.'), function noop() {});
+    sendReport(new Error('Timeout Exceeded.'));
   }, endTime);
 }
 
-module.exports = options => {
-  const fn = func => {
-    var dnsPromise = undefined;
-    fn.metricsQueue = [];
-    const config = setConfig(options);
-
-    if (!config.clientId) {
-      // No-op if user doesn't set an IOpipe token.
-      return func;
+class IOpipeWrapperClass {
+  constructor(
+    dnsPromise,
+    config,
+    userFunc,
+    originalEvent,
+    originalContext,
+    originalCallback
+  ) {
+    this.startTime = process.hrtime();
+    this.config = config;
+    this.metrics = [];
+
+    // assign a new dnsPromise if it's not a coldstart because dns could have changed
+    if (!globals.COLDSTART) {
+      this.dnsPromise = getDnsPromise(this.config.host);
+    } else {
+      this.dnsPromise = dnsPromise;
     }
 
-    /* resolve DNS early on coldstarts */
-    dnsPromise = makeDnsPromise(config.host);
+    this.originalContext = originalContext;
+    this.originalCallback = originalCallback;
 
-    return function IOpipeWrapper() {
-      fn.metricsQueue = [];
-      const args = [].slice.call(arguments);
-
-      if (!globals.COLDSTART) {
-        /* Get an updated DNS record. */
-        dnsPromise = makeDnsPromise(config.host);
-      }
-
-      const startTime = process.hrtime();
-      const report = new Report(
-        config,
-        args[1],
-        startTime,
-        fn.metricsQueue,
-        dnsPromise
+    // preserve original functions via a property name change
+    ['succeed', 'fail', 'done'].forEach(method => {
+      Object.defineProperty(
+        this.originalContext,
+        `original_${method}`,
+        Object.getOwnPropertyDescriptor(this.originalContext, method)
       );
+    });
 
-      const timeout = setupTimeoutCapture(config, report, args[1]);
-
-      const callback = (err, cb) => {
-        if (timeout) {
-          clearTimeout(timeout);
-        }
-        report.send(err, cb);
-      };
-
-      /* Mangle arguments, wrapping callbacks. */
-      args[1] = Context(callback, args[1]);
-      args[2] = Callback(callback, args[2]);
-
-      try {
-        return func.apply(this, args);
-      } catch (err) {
-        clearTimeout(timeout);
-        report.send(err, function noop() {});
-        return undefined;
+    // assign modified methods and objects here
+    this.modifiedContext = Object.assign(this.originalContext, {
+      // need to use .bind, otherwise, the this ref inside of each fn is NOT IOpipeWrapperClass
+      succeed: this.succeed.bind(this),
+      fail: this.fail.bind(this),
+      done: this.done.bind(this),
+      iopipe: {
+        log: this.log.bind(this),
+        metrics: this.metrics,
```

Comment:
I don't think we should expose this. It makes it easy for users to put in invalid data. We filter it anyway, but it's an interface I'm not keen on adding? Getters/setters would be interesting though, if we wanted an assignment mechanism instead of the .log()